sycl/doc/extensions/experimental/sycl_ext_oneapi_graph.asciidoc

```diff
@@ -1548,6 +1690,167 @@ different from the one with which the dynamic_parameter was created.
+submitted to a SYCL queue. External graph events are associated with a single
+external event node inside a graph. External events are functionally normal SYCL
+events with some restrictions (see below).
+
+External events can be obtained by either:
+
+* Calling `command_graph::add_external_event()` on a modifiable `command_graph`
+to add an external event node, then obtain the SYCL event object by calling
+`node::get_external_event()`. 
+
+* Calling `queue::ext_oneapi_external_event()` on a queue in the recording
+state, which will return an external SYCL event.
+
+External events may be used as parameters to `handler::depends_on()` or as
+dependent events for queue submissions outside of the graph they
+are associated with. For example, this allows enqueing work outside of the graph
+part-way through the graph execution, rather than having to wait for the entire
+graph to finish.
+
+External events are considered complete when all their dependent graph nodes
+have finished execution, and their execution status is automatically reset when
+the executable graph which contains them is submitted to a queue for execution.
+
+External events have the following restrictions:
+
+* Profiling is not currently supported for external graph events and calling
+`event::get_profiling_info()` will throw synchronously with error code
+`invalid`.
+
+* External events cannot be passed as dependent events for submissions within
+the same graph, and doing so will throw synchronously with error code `invalid`.
+Graph-limited events should be used for this purpose.
+
+[source,c++]
+----
+/* Basic usage example using the Record&Replay API. */
+
+// Create the graph `A`.
+sycl::queue queueA;
+sycl_ext::command_graph graphA(queueA.get_context(), queueA.get_device());
+
+// Start recording the queue.
+graphA.begin_recording(queueA);
+
+// Submit kernel on the recorded queue.
+// EventA is a graph-limited event that can only be used for managing
+// graph internal dependencies.
+EventA = queueA.submit([&](handler& cgh) {...});
+
+event externalEventA = queueA.ext_oneapi_external_event({EventA});
+
+// Stop recording the queue.
+graphA.end_recording();
+
+// Create another graph `B`
+sycl::queue queueB;
+sycl_ext::command_graph graphB(queueB.get_context(), queueB.get_device());
+
+// Start recording the queue.
+graphB.begin_recording(queueB);
+
+// Submit kernel on the recorded queue.
+queueB.submit([&](handler& cgh) {...});
+
+queueB.ext_oneapi_submit_barrier({externalEventA});
+
+// Stop recording the queue.
+graphB.end_recording();
+
+// Finalize the modifiable graphs to create an executable graphs that can be
+// submitted for execution.
+auto execGraphA = graphA.finalize();
+auto execGraphB = graphB.finalize();
+
+// Execute graph
+for(int i = 0 ; i < Iterations, i++) {
+  queueA.submit([&](handler& cgh) {
+    cgh.ext_oneapi_graph(execGraphA);
+  })
+
+  queueB.submit([&](handler& cgh) {
+    cgh.ext_oneapi_graph(execGraphB);
+  })
```

Comment:
what will happen if we first submit execGraphB and then submit execGraphA,   endless wait within execGraphB?

Reply:
Since they are separate graphs with no implicit dependencies from accessors between them, then no I believe that execGraphA could still execute after execGraphB. It may endlessly wait if there were implicit dependencies though I think.

Reply:
not quite understand that 'It **may** endlessly wait **if there were implicit dependencies**'.

I agree that execGraphA could run successfully since it does not depend on others.

While for execGraphB, there's a clear explicit dependency that it depends on externalEventA, not "if there were implicit dependencies", so my point is that it always endlessly waits if launch first, not "may endlessly wait".

Reply:
So the external events should be implemented as a device-side dependency, so `execGraphB` will not block on the host waiting for the external event because there is no host synchronization there. Since `execGraphB` does not block on host that allows the application to proceed and execute `execGraphA`, and once it reaches the point where the external event is signaled then `execGraphB` (on the device) can also proceed to finish execution. Does that make sense?

My point about implicit dependencies is more of a runtime implementation detail I suppose, but in SYCL using accessors creates implicit dependencies between commands (and we also extend this to graphs), so if you were using accessors which created a dependency this may cause a blocking host wait on attempting to execute `execGraphA` which would deadlock in this case.

This could potentially be mitigated by specifying exactly when/how the external event status changes. For example, if we consider the external event to begin in the complete state and be reset upon the owning graph being executed then that would prevent this scenario from being a deadlock, at least.

Reply:
> So the external events should be implemented as a device-side dependency, so `execGraphB` will not block on the host waiting for the external event because there is no host synchronization there. Since `execGraphB` does not block on host that allows the application to proceed and execute `execGraphA`, and once it reaches the point where the external event is signaled then `execGraphB` (on the device) can also proceed to finish execution. Does that make sense?

yes, for this case, `execGraphB` waits on the device side until there is a host function changes the event status to complete.

> 
> My point about implicit dependencies is more of a runtime implementation detail I suppose, but in SYCL using accessors creates implicit dependencies between commands (and we also extend this to graphs), so if you were using accessors which created a dependency this may cause a blocking host wait on attempting to execute `execGraphA` which would deadlock in this case.

just wonder, for most popular cases, the buffer is always accessed by the device, and there is no internal D2H or H2D copy, so, my understanding is that the wait also happens at the device side, why the host side?

> 
> This could potentially be mitigated by specifying exactly when/how the external event status changes. For example, if we consider the external event to begin in the complete state and be reset upon the owning graph being executed then that would prevent this scenario from being a deadlock, at least.

yes, good point.

Reply:
> just wonder, for most popular cases, the buffer is always accessed by the device, and there is no internal D2H or H2D copy, so, my understanding is that the wait also happens at the device side, why the host side?

You are right actually this should be a device side wait in all cases, I was remembering old workarounds in the backend.